### PR TITLE
lu: When not using pivoting, return the identity permutation instead of zeros

### PR DESCRIPTION
**Pull request by bamos**
Some of my qpth users have told me that updating to the latest version of PyTorch and replacing the btrifact/btrisolve calls with the LU ones wasn't working and I didn't believe them until I tried it myself :)

These updates have broken unpivoted LU factorizations/solves on CUDA. The LU factorization code used to return the identity permutation when pivoting wasn't used but now returns all zeros as the pivots. This PR reverts it back to return the identity permutation. I've not yet tested this code as I'm having some trouble compiling PyTorch with this and am hitting #21700 and am not sure how to disable that option.

Here's a MWE to reproduce the broken behavior, and my fix.

```python
torch.manual_seed(0)

n = 4
L = torch.randn(n,n)
A = L.mm(L.t()).unsqueeze(0)
b = torch.randn(1, n)

A_lu_cpu = torch.lu(A)
A_lu_cuda_nopivot = torch.lu(A.cuda(), pivot=False)
A_lu_cuda_pivot = torch.lu(A.cuda(), pivot=True)
print('A_lu_cuda_nopivot\n', A_lu_cuda_nopivot)
print('-----\nA_lu_cuda_pivot\n', A_lu_cuda_nopivot)

x_cpu = b.lu_solve(*A_lu_cpu)
x_cuda_nopivot = b.cuda().lu_solve(*A_lu_cuda_nopivot)
x_cuda_nopivot_fixed = b.cuda().lu_solve(
    A_lu_cuda_nopivot[0], torch.arange(1, n+1, device='cuda:0').int())
x_cuda_pivot = b.cuda().lu_solve(*A_lu_cuda_pivot)

print(x_cpu, x_cuda_nopivot, x_cuda_nopivot_fixed, x_cuda_pivot)
```

Output:

```
A_lu_cuda_nopivot
 (tensor([[[ 2.8465, -0.7560,  0.8716, -1.7337],
         [-0.2656,  5.5724, -1.1316,  0.6678],
         [ 0.3062, -0.2031,  1.4206, -0.5438],
         [-0.6091,  0.1198, -0.3828,  1.5103]]], device='cuda:0'), tensor([[0, 0, 0, 0]], device='cuda:0', dtype=torch.int32))


-----


A_lu_cuda_pivot
 (tensor([[[ 2.8465, -0.7560,  0.8716, -1.7337],
         [-0.2656,  5.5724, -1.1316,  0.6678],
         [ 0.3062, -0.2031,  1.4206, -0.5438],
         [-0.6091,  0.1198, -0.3828,  1.5103]]], device='cuda:0'), tensor([[0, 0, 0, 0]], device='cuda:0', dtype=torch.int32))


(tensor([[-0.3121, -0.1673, -0.4450, -0.2483]]),
 tensor([[-0.1661, -0.1875, -0.5694, -0.4772]], device='cuda:0'),
 tensor([[-0.3121, -0.1673, -0.4450, -0.2483]], device='cuda:0'),
 tensor([[-0.3121, -0.1673, -0.4450, -0.2483]], device='cuda:0'))
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu
Comment (vishwakftw) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: req_size includes batch dimensions, and hence this will fail on tensors with greater than 4 dimensions.

This snippet should work fine:
```
Tensor pivots_tensor = at::arange(1, self.size(-1) + 1, self.options().dtype(at::kInt)).expand_as(req_size);
```

Also, a test would be nice.

**Review by vishwakftw — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu
Comment (bamos) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: I've updated to that, and have added tests. Still untested on my machine because other build issues.

Comment (vishwakftw) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: You can set USE_MKLDNN=0 and try building if #21700 is bothering you :)

Comment (bamos) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: I quickly tried that yesterday and cleaned+rebuilt but still hit that issue

Comment (bamos) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: Also why would the LU factorizations/solves be called on tensors with 4 or more dimensions? In the batched case, `A` is `[n_batch, n, m]` and `b` is ~`[n_batch, m, n_rhs]`, right?

Comment (bamos) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: (Maybe yesterday I didn't properly clean my PyTorch build when trying to disable that)

Comment (vishwakftw) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: 3 dimensions would incur a reshaping cost if the input tensor is > 3 dimensions.

Comment (vishwakftw) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: @bamos I'm sorry. It should've been `expand` instead of `expand_as`, and that has caused the failures.

Comment (bamos) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: Ah sorry for trying to send this in without a working build, just updated that

Comment (vishwakftw) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: Thank you, I'll wait for CI to finish.

**Review by vishwakftw — DISMISSED** [review]
Thank you for fixing this. I’ll pre-approve this for now.

File: test/test_torch.py
Comment (vishwakftw) [inline test/test_torch.py]: Shouldn’t this be changed too?

Comment (bamos) [reply test/test_torch.py]: Updated

File: test/test_torch.py
Comment (vishwakftw) [inline test/test_torch.py]: Nit: a.size(-1)

Comment (bamos) [reply test/test_torch.py]: updated

Comment (vishwakftw) [reply test/test_torch.py]: One last nit: sorry about the inconvenience.

```
self.assertEqual(nopiv, cast(torch.arange(1, 1 + a.size(-1), dtype=torch.int32).expand(a.shape[:-1])))
```

Comment (bamos) [reply test/test_torch.py]: added

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu
Comment (vishwakftw) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: I think we need to expand and clone rather than just expand. `lu_solve` requires `pivots` to be contiguous, and expanding this tensor to higher dimensions makes it non-contiguous. Additionally, this is also used in the MAGMA routines, so having expanded tensors will cause memory to overwritten.

File: test/test_torch.py
Comment (vishwakftw) [inline test/test_torch.py]: As the name suggests, `pivot` should be `False` here always. This is also causing test failures.

**Review by vishwakftw — COMMENTED** [review]
After investigating the test failures, I've localized the issues. I am 99% sure that the tests would pass now. I'm sorry about reviewing again and again and keeping the PR in a state of flux.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu
Comment (bamos) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: Hmm, would it be simpler if you just put these changes directly into my fork for? I just added you as a collaborator there so you should be able to push to this branch

Comment (vishwakftw) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: Great, thanks!

**Review by vishwakftw — APPROVED** [review]
The failing Windows test has passed which most probably indicates that the patch is correct. Should be good to merge after CI finishes. Thank you @bamos for the PR!

**Comment by vishwakftw** [comment]
Failures seem to unrelated.

@pytorchbot merge this please

**Comment by vishwakftw** [comment]
@ezyang is this good to go?

**Pull request closed without merging**
